tests/unit/s2n_signature_algorithms_test.c

```diff
@@ -162,78 +162,159 @@ int main(int argc, char **argv)
+                EXPECT_OK(s2n_signature_algorithm_recv(conn, &empty));
+                EXPECT_EQUAL(conn->handshake_params.server_cert_sig_scheme, &s2n_ecdsa_sha1);
+            };
+
+            /* Server chooses default based on client cert type */
+            {
+                DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_SERVER),
+                        s2n_connection_ptr_free);
+                conn->security_policy_override = &test_security_policy;
+                conn->actual_protocol_version = S2N_TLS11;
+
+                conn->handshake_params.client_cert_pkey_type = S2N_PKEY_TYPE_RSA;
+                EXPECT_OK(s2n_signature_algorithm_recv(conn, &empty));
+                EXPECT_EQUAL(conn->handshake_params.client_cert_sig_scheme, &s2n_rsa_pkcs1_md5_sha1);
+
+                conn->handshake_params.client_cert_pkey_type = S2N_PKEY_TYPE_ECDSA;
+                EXPECT_OK(s2n_signature_algorithm_recv(conn, &empty));
+                EXPECT_EQUAL(conn->handshake_params.client_cert_sig_scheme, &s2n_ecdsa_sha1);
+            };
         };
 
-        /* Test: don't negotiate invalid signatures (protocol not high enough) */
+        /* Test: don't negotiate signature scheme not allowed by security policy */
         {
-            const struct s2n_signature_scheme *result = NULL;
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            conn->security_policy_override = &test_security_policy;
 
-            s2n_stuffer_wipe(&choice);
-            s2n_stuffer_write_uint16(&choice, s2n_ecdsa_secp384r1_sha384.iana_value);
+            DEFER_CLEANUP(struct s2n_stuffer input = { 0 }, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&input, 0));
 
-            conn->actual_protocol_version = S2N_TLS13;
-            EXPECT_SUCCESS(s2n_get_and_validate_negotiated_signature_scheme(conn, &choice, &result));
-            EXPECT_EQUAL(result, &s2n_ecdsa_secp384r1_sha384);
+            const struct s2n_signature_scheme *unsupported = &s2n_rsa_pkcs1_sha512;
+            const struct s2n_signature_preferences *prefs = test_security_policy.signature_preferences;
+            for (size_t i = 0; i < prefs->count; i++) {
+                EXPECT_NOT_EQUAL(prefs->signature_schemes[i]->iana_value,
+                        unsupported->iana_value);
+            }
 
-            s2n_stuffer_reread(&choice);
-            conn->actual_protocol_version = S2N_TLS12;
-            EXPECT_FAILURE_WITH_ERRNO(s2n_get_and_validate_negotiated_signature_scheme(conn, &choice, &result),
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&input, unsupported->iana_value));
+            EXPECT_ERROR_WITH_ERRNO(s2n_signature_algorithm_recv(conn, &input),
                     S2N_ERR_INVALID_SIGNATURE_SCHEME);
         };
 
+        /* Test: don't negotiate default signature scheme not allowed by security policy */
```

Comment:
This was specifically ensuring that for TLS1.2 we rejected a signature algorithm chosen by the peer that wasn't in our security policy, and didn't make some special exception for the legacy defaults. Because we've made that mistake before :)

But it was mostly a duplicate of the previous test but with different test signature schemes, so I combined them, added a control / sanity check, and tried to clean up the comment. See if that's better.

Reply:
That makes sense, thanks! The new combined test looks good.